### PR TITLE
"Game List" Mode Overhaul: Phase 2

### PR DESCRIPTION
**Pull request by RetraCarteR**
In **Pull Request #105**, I outlined a three-phase plan to improve the ROM browser's "Game List" mode. That pull request, which implemented phase one of this plan, was merged into RMG and made accessible to the masses with the release of version 0.3.6. 

As of right now, I'm proud to present: **Phase 2: Hide-able Columns**. Be sure to read the description of the included commit for details as to **a)** how this new feature works, and **b)** a few other minor changes I've made to how "Game List" mode functions.

Here are a few pictures of the new hide-able column feature in action:
RMG_1
RMG_2

I eagerly await your review!

### CODE REVIEW

**Comment by RetraCarteR** [comment]
Picture of the main context menu as of my latest commit:
RMG_5

**Comment by RetraCarteR** [comment]
Picture of the main context menu as of my latest commit:
RMG_6

File: Source/RMG-Core/Settings/Settings.cpp [hunk 1]
```diff
@@ -547,6 +547,9 @@ static l_Setting get_setting(SettingsID settingId)
     case SettingsID::RomBrowser_ColumnSizes:
         setting = {SETTING_SECTION_ROMBROWSER, "ColumnSizes", std::vector<int>({-1, -1, -1})};
         break;
+    case SettingsID::RomBrowser_ColumnVisibility:
```
Comment (Rosalie241) [inline Source/RMG-Core/Settings/Settings.cpp:550]: Why not just call thix `RomBrowser_Columns`? It'd make sense for yor third step anyways, just have more columns but just having some hidden by default.

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp [hunk 2]
```diff
@@ -95,13 +95,13 @@ RomBrowserWidget::RomBrowserWidget(QWidget *parent) : QStackedWidget(parent)
     this->listViewWidget->horizontalHeader()->setDefaultAlignment(Qt::AlignLeft);
     this->listViewWidget->horizontalHeader()->setSortIndicatorShown(false);
     this->listViewWidget->horizontalHeader()->setHighlightSections(false);
-    this->listViewWidget->horizontalHeader()->setContextMenuPolicy(Qt::ContextMenuPolicy::CustomContextMenu);
+    this->listViewWidget->horizontalHeader()->setContextMenuPolicy(Qt::CustomContextMenu);
```
Comment (Rosalie241) [inline Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:98]: Please revert `Qt::CustomContextMenu` back to `Qt::ContextMenuPolicy::CustomContextMenu);`

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp [hunk 3]
```diff
@@ -95,13 +95,13 @@ RomBrowserWidget::RomBrowserWidget(QWidget *parent) : QStackedWidget(parent)
...
     this->addWidget(this->listViewWidget);
     connect(this->listViewWidget, &QTableView::doubleClicked, this, &RomBrowserWidget::on_DoubleClicked);
-    connect(this->listViewWidget, &QTableView::customContextMenuRequested, this, &RomBrowserListViewWidget::customContextMenuRequested);
```
Comment (Rosalie241) [inline Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:101]: Why was this changed to connecting to the stacked widget's custom context menu's signal?

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp [hunk 4]
```diff
@@ -95,13 +95,13 @@ RomBrowserWidget::RomBrowserWidget(QWidget *parent) : QStackedWidget(parent)
...
     connect(this->listViewWidget->horizontalHeader(), &QHeaderView::sortIndicatorChanged, this, &RomBrowserWidget::on_listViewWidget_sortIndicatorChanged);
     connect(this->listViewWidget->horizontalHeader(), &QHeaderView::sectionResized, this, &RomBrowserWidget::on_listViewWidget_sectionResized);
     connect(this->listViewWidget->horizontalHeader(), &QHeaderView::sectionMoved, this, &RomBrowserWidget::on_listViewWidget_sectionMoved);
+    connect(this->listViewWidget->horizontalHeader(), &QHeaderView::customContextMenuRequested, this, &RomBrowserWidget::on_listViewWidget_headerContextMenuRequested);
```
Comment (Rosalie241) [inline Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:104]: Instead of making a new signal slot and connecting to it, you can connect oto the main custom context menu function.

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp [hunk 5]
```diff
@@ -128,23 +128,26 @@ RomBrowserWidget::RomBrowserWidget(QWidget *parent) : QStackedWidget(parent)
...
-    this->setContextMenuPolicy(Qt::ContextMenuPolicy::CustomContextMenu);
-    connect(this, &QStackedWidget::customContextMenuRequested, this, &RomBrowserWidget::customContextMenuRequested);
+    this->setContextMenuPolicy(Qt::CustomContextMenu);
+    connect(this, &QStackedWidget::customContextMenuRequested, this, &RomBrowserWidget::mainContextMenuRequested);
 
-    // configure context menu actions
+    // define context menus
+    this->mainContextMenu = new QMenu(this);
+    this->headerContextMenu = new QMenu(this);
+    this->categorySubmenu = new QMenu(this);
```
Comment (Rosalie241) [inline Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp]: It's not a ctegory submenu so please name this ike `menu_Columns` and add it to the bottom of the context menu action definitions

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp [hunk 6]
```diff
@@ -128,23 +128,26 @@ RomBrowserWidget::RomBrowserWidget(QWidget *parent) : QStackedWidget(parent)
...
     // configure context menu policy
-    this->setContextMenuPolicy(Qt::ContextMenuPolicy::CustomContextMenu);
-    connect(this, &QStackedWidget::customContextMenuRequested, this, &RomBrowserWidget::customContextMenuRequested);
+    this->setContextMenuPolicy(Qt::CustomContextMenu);
+    connect(this, &QStackedWidget::customContextMenuRequested, this, &RomBrowserWidget::mainContextMenuRequested);
 
-    // configure context menu actions
+    // define context menus
+    this->mainContextMenu = new QMenu(this);
+    this->headerContextMenu = new QMenu(this);
```
Comment (Rosalie241) [inline Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:147]: This is unrequired

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp [hunk 7]
```diff
@@ -128,23 +128,26 @@ RomBrowserWidget::RomBrowserWidget(QWidget *parent) : QStackedWidget(parent)
...
 
     // configure context menu policy
-    this->setContextMenuPolicy(Qt::ContextMenuPolicy::CustomContextMenu);
-    connect(this, &QStackedWidget::customContextMenuRequested, this, &RomBrowserWidget::customContextMenuRequested);
+    this->setContextMenuPolicy(Qt::CustomContextMenu);
+    connect(this, &QStackedWidget::customContextMenuRequested, this, &RomBrowserWidget::mainContextMenuRequested);
 
-    // configure context menu actions
+    // define context menus
+    this->mainContextMenu = new QMenu(this);
```
Comment (Rosalie241) [inline Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:146]: Please revert the name change

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp [hunk 8]
```diff
@@ -176,24 +182,6 @@ RomBrowserWidget::RomBrowserWidget(QWidget *parent) : QStackedWidget(parent)
     connect(this->action_SetCoverImage, &QAction::triggered, this, &RomBrowserWidget::on_Action_SetCoverImage);
     connect(this->action_RemoveCoverImage, &QAction::triggered, this, &RomBrowserWidget::on_Action_RemoveCoverImage);
 
-    // configure context menu
```
Comment (Rosalie241) [inline Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:179]: Please revert the changes here, clearing and re-adding the actions in the context menu request slot is unneeded and kinda wasteful

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp [hunk 9]
```diff
@@ -446,7 +434,7 @@ void RomBrowserWidget::on_DoubleClicked(const QModelIndex& index)
     emit this->PlayGame(this->getCurrentRom());
 }
 
-void RomBrowserWidget::customContextMenuRequested(QPoint position)
+void RomBrowserWidget::mainContextMenuRequested(QPoint position)
```
Comment (Rosalie241) [inline Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:437]: Rever the name change please

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp [hunk 10]
```diff
@@ -460,15 +448,27 @@ void RomBrowserWidget::customContextMenuRequested(QPoint position)
 
     this->getCurrentData(data);
 
+    this->mainContextMenu->clear();
```
Comment (Rosalie241) [inline Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:451]: Like I mentioned above, please revert this

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp [hunk 11]
```diff
@@ -460,15 +448,27 @@ void RomBrowserWidget::customContextMenuRequested(QPoint position)
...
+
     this->action_PlayGame->setEnabled(hasSelection);
     this->action_PlayGameWith->setEnabled(hasSelection);
     this->action_RomInformation->setEnabled(hasSelection);
     this->action_EditGameSettings->setEnabled(hasSelection);
     this->action_EditCheats->setEnabled(hasSelection);
-    this->action_SetCoverImage->setEnabled(hasSelection);
-    this->action_SetCoverImage->setVisible(view == this->gridViewWidget);
-    this->action_RemoveCoverImage->setEnabled(hasSelection && !data.coverFile.isEmpty());
-    this->action_RemoveCoverImage->setVisible(view == this->gridViewWidget);
```
Comment (Rosalie241) [inline Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:471]: Please also revert this change and add the new columns qmenu below this in the same way as it was done here before your removal.

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp [hunk 12]
```diff
@@ -479,8 +479,19 @@ void RomBrowserWidget::customContextMenuRequested(QPoint position)
         this->action_PlayGameWith->setText("Play Game with Disk");
     }
 
+    if (view == this->listViewWidget)
+    { // list view
+        this->categorySubmenu = this->mainContextMenu->addMenu(tr("&Show/Hide Columns"));
```
Comment (Rosalie241) [inline Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp]: 1) why the `&`
2) please remove the `tr` usage, RMG wont support translations anytime soon

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp [hunk 13]
```diff
@@ -535,6 +556,32 @@ void RomBrowserWidget::on_listViewWidget_sectionMoved(int logicalIndex, int oldV
     CoreSettingsSetValue(SettingsID::RomBrowser_ColumnOrder, columnOrder);
 }
 
+void RomBrowserWidget::on_listViewWidget_headerContextMenuRequested(QPoint position)
```
Comment (Rosalie241) [inline Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:559]: This function should be removed

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp [hunk 14]
```diff
@@ -535,6 +556,32 @@ void RomBrowserWidget::on_listViewWidget_sectionMoved(int logicalIndex, int oldV
...
+{
+    this->headerContextMenu->clear();
+    RomBrowserWidget::on_listViewWidget_columnVisibilityMenuRequested(headerContextMenu);
+    this->headerContextMenu->popup(this->mapToGlobal(position));
+}
+
+void RomBrowserWidget::on_listViewWidget_columnVisibilityMenuRequested(QMenu* currentMenu)
```
Comment (Rosalie241) [inline Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:566]: This is not a real signal slot function so it shouldn't be named as such, maybe a name like `addColumnsToMenu` would be better.

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp [hunk 15]
```diff
@@ -479,8 +479,19 @@ void RomBrowserWidget::customContextMenuRequested(QPoint position)
         this->action_PlayGameWith->setText("Play Game with Disk");
     }
 
+    if (view == this->listViewWidget)
+    { // list view
+        this->categorySubmenu = this->mainContextMenu->addMenu(tr("&Show/Hide Columns"));
+        RomBrowserWidget::on_listViewWidget_columnVisibilityMenuRequested(categorySubmenu);
```
Comment (Rosalie241) [inline Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp]: Please just use `this->functionName()` instead of the odd `classname::functionName();`

**Review by Rosalie241 — CHANGES_REQUESTED** [review]

**Comment by RetraCarteR** [comment]
I'll get to work on those changes when I have the time. I'll let you know on Discord once everything's been addressed.

File: Source/RMG-Core/Settings/Settings.cpp (hunk 1 above)
Comment (RetraCarteR) [reply Source/RMG-Core/Settings/Settings.cpp:550]: Done, will include in next commit.

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp (hunk 2 above)
Comment (RetraCarteR) [reply Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:98]: Done, will include in next commit.

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp (hunk 5 above)
Comment (RetraCarteR) [reply Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp]: Done, will include in next commit (I actually already renamed it in an additional commit I posted while you were doing your review but before you posted it).

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp (hunk 7 above)
Comment (RetraCarteR) [reply Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:146]: Done, will include in next commit.

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp (hunk 9 above)
Comment (RetraCarteR) [reply Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:437]: Done, will include in next commit.

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp (hunk 10 above)
Comment (RetraCarteR) [reply Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:451]: Done, will include in next commit.

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp (hunk 15 above)
Comment (RetraCarteR) [reply Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp]: Done, will include in next commit. (Don't know how this one slipped by me.)

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp (hunk 4 above)
Comment (RetraCarteR) [reply Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:104]: Done, will include in next commit. (My original plan was to bring up just the column menu when the header is right-clicked. This change makes it so that doing so pulls up the full context menu instead. Is that what you want? I'd just like to make sure.)

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp (hunk 13 above)
Comment (RetraCarteR) [reply Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:559]: Done, will include in next commit. (See my comment on your fourth requested change).

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp (hunk 3 above)
Comment (RetraCarteR) [reply Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:101]: My understanding was that the stacked widget encompasses all four ROM browser modes and that the "customContextMenuRequested" function already sets it to only appear when the ROM browser is in "Game List" or "Game Grid" modes. If that is the case, then having separate context menu signals for each mode should be redundant, right? Let me know if I'm wrong.

Using just the stacked widget's signal for everything didn't seem to cause any issues in my testing, but I can re-add the individual signals if you want me to or if I'm wrong about it being unnecessary.

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp (hunk 6 above)
Comment (RetraCarteR) [reply Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:147]: Are you sure? Deleting the context menu definitions causes RMG to crash for me whenever the context menu is opened.

Comment (Rosalie241) [reply Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:147]: After my overlook, this is now fine, I'll mark it as resolved!

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp (hunk 11 above)
Comment (RetraCarteR) [reply Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:471]: Done, will include in next commit. (I implemented the submenu this way because I was having trouble setting its visibility. I was trying to call "setVisible()" on the submenu itself when I should have been calling it on the submenu's "menuAction()". Now that I've figured that out, I've implemented the submenu the same way as all of the other actions in the context menu.)

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp (hunk 8 above)
Comment (RetraCarteR) [reply Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:179]: Done, will include in next commit. (In order for the "Show/Hide Columns" submenu to keep track of column order and update itself accordingly, the columns submenu has to be rebuilt every time it's pulled up. I didn't know how to clear individual actions, so _that's_ why I set the entire context menu to clear and rebuild itself every time. I've since figured out how to reset the submenu by itself, and I've updated the code accordingly.)

File: Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp (hunk 14 above)
Comment (RetraCarteR) [reply Source/RMG/UserInterface/Widget/RomBrowserWidget.cpp:566]: Done, will include in next commit. (It's been renamed to "generateColumnsMenu").
